CRASH (any WOW64 app) on AMD Ryzen in drlibc_x86.asm:566 WOW64 mode switch

**Describe the bug**

bin32\ddrun.exe crashes most of the time when operating on 32 bit applications. Maybe one out of 100 times notepad.exe launches. In my setup, Windows 10 is being virtualized by KVM. AMD 3900x Processor is being used.  Older versions of DynamoRio has the same issue.  bin64\drrun.exe works fine for 64 bit applications.

```
C:\Users\justin\Downloads\DynamoRIO-Windows-7.91.18301-1\DynamoRIO-Windows-7.91.18301-1>bin32\drrun.exe -- C:\windows\syswow64\notepad.exe
```

![image](https://user-images.githubusercontent.com/8340498/74109412-38775400-4b51-11ea-94e8-271694ac1a3a.png)


**To Reproduce**
Steps to reproduce the behavior:
1. Download Window's cronbuild-7.91.18301A from https://github.com/DynamoRIO/dynamorio/releases . 
2. Extract and cd to DynamoRIO folder.
3. Run the command `bin32\drrun.exe -- C:\windows\syswow64\notepad.exe`
4. Note the application likely crashes.
5. Repeat step 3  around 100 times, and notice notepad launches.

Please also answer these questions:
 - What happens when you run without any client?

No client is being used. Clients do work, but again need to run 100 times or so to get the client to run. I

 - What happens when you run with debug build ("-debug" flag to drrun/drconfig/drinject)?

Same issue occurs, but with more detailed error message. When notepad manages to start, it takes a longer time for the process to open.

Image showing the error message:

![image](https://user-images.githubusercontent.com/8340498/74109498-ebe04880-4b51-11ea-9bf7-8817ca4a4b04.png)

Image showing notepad.exe does launch once in a while:

![image](https://user-images.githubusercontent.com/8340498/74109515-1f22d780-4b52-11ea-9cf2-ac2722c341b0.png)


**Expected behavior**

Notepad to open and run.

**Screenshots or Pasted Text**

Following is StackTrace in WinDBG. The above reproduction steps is for starting in windows without
```
ModLoad: 75ab0000 75ad5000   C:\Windows\SysWOW64\IMM32.DLL
ModLoad: 14460000 14909000   C:\Users\justin\Downloads\DynamoRIO-Windows-7.91.18301-1\DynamoRIO-Windows-7.91.18301-1\lib32\debug\dynamorio.dll
(1348.2a8): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
eax=00000000 ebx=00fbf490 ecx=b58d14d4 edx=00000000 esi=2190b860 edi=00fbf540
eip=14775eab esp=00fbf4b0 ebp=00fbf4e4 iopl=0         nv up ei pl nz ac po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010212
dynamorio!switch_modes_and_call+0x4e:
14775eab 5b              pop     ebx
1:002> k
 # ChildEBP RetAddr  
00 00fbf4e4 146fe909 dynamorio!switch_modes_and_call+0x4e [C:\projects\dynamorio\build\build_debug-32\core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj.s @ 1961] 
01 00fbf510 146fed98 dynamorio!os_take_over_wow64_extra+0x49 [c:\projects\dynamorio\core\win32\os.c @ 2433] 
02 00fbfa30 146f1928 dynamorio!os_take_over_thread+0x2b8 [c:\projects\dynamorio\core\win32\os.c @ 2571] 
03 00fbfa70 1446fc58 dynamorio!os_take_over_all_unknown_threads+0x1f8 [c:\projects\dynamorio\core\win32\os.c @ 2707] 
04 00fbfaa4 146d7322 dynamorio!dynamorio_take_over_threads+0xf8 [c:\projects\dynamorio\core\dynamo.c @ 2904] 
05 00fbfb10 146d7aa8 dynamorio!auto_setup+0x202 [c:\projects\dynamorio\core\arch\x86_code.c @ 198] 
06 00000000 00000000 dynamorio!dynamo_auto_start+0x8 [C:\projects\dynamorio\build\build_debug-32\core\CMakeFiles\dynamorio.dir\arch\x86\x86.asm.obj.s @ 1600] 
```

```
1:003> .exr -1
ExceptionAddress: 70afe90b (dynamorio!switch_modes_and_call+0x0000004e)
   ExceptionCode: c0000005 (Access violation)
  ExceptionFlags: 00000000
NumberParameters: 2
   Parameter[0]: 00000003
   Parameter[1]: 00000000
Attempt to write to address 00000000
```
```
0:000> !address esp

                                     
Mapping file section regions...
Mapping module regions...
Mapping PEB regions...
Mapping TEB and stack regions...
Mapping heap regions...
Mapping page heap regions...
Mapping other regions...
Mapping stack trace database regions...
Mapping activation context regions...

Usage:                  Stack
Base Address:           00fff000
End Address:            01010000
Region Size:            00011000 (  68.000 kB)
State:                  00001000          MEM_COMMIT
Protect:                00000004          PAGE_READWRITE
Type:                   00020000          MEM_PRIVATE
Allocation Base:        00fd0000
Allocation Protect:     00000004          PAGE_READWRITE
More info:              ~0k


Content source: 1 (target), length: bd0
```

**Versions**
 - What version of DynamoRIO are you using?
Windows-7.91.18301

 - Does the latest build from
https://github.com/DynamoRIO/dynamorio/wiki/Latest-Build solve the problem?
No

- What operating system version are you running on? ("Windows 10" is *not* sufficient: give the release number.)

Windows 10 1809 (Build 17763.973) is being used. I am virtualizing  Windows through  KVM.

- Is your application 32-bit or 64-bit?

32 bit

**Additional context**
None

Sorry I might need more help. I'm not very fluent with Windows and Visual Studio.

I've added "DECL_EXTERN(os_thread_yield)" and put in a "call os_thread_yield", but I get linking failures in various projects, e.g.:

```
Creating library C:/Users/Worry/projects/dynamorio/build/lib32/drinjectlib.lib and object C:/Users/Worry/projects/dynamorio/build/lib32/drinjectlib.exp
drlibc.lib(drlibc_x86.asm.obj) : error LNK2019: unresolved external symbol _os_thread_yield referenced in function _switch_modes_and_call [C:\Users\Worry\projects\dynamorio\build\core\drinjectlib.vcxproj]
C:\Users\Worry\projects\dynamorio\build\lib32\drinjectlib.dll : fatal error LNK1120: 1 unresolved externals [C:\Users\Worry\projects\dynamorio\build\core\drinjectlib.vcxproj]
```
I need to add the COFF file os.obj to various failing projects? I'm not sure of the VS solution method of doing that.

--- a/core/CMakeLists.txt
+++ b/core/CMakeLists.txt
@@ -106,6 +106,8 @@ set(DRLIBC_SRCS
   ${drlibc_arch_asm_src}
   drlibc/drlibc.c
   drlibc/drlibc_notdr_dcxt.c
+  drlibc/drlibc_notdr_error.c
+  drlibc/drlibc_notdr_ignoreassert.c
   drlibc/drlibc_notdr_logfile.c
   drlibc/drlibc_notdr_printlog.c
   drlibc/drlibc_notdr_report.c

--- a/core/drlibc/drlibc.c
+++ b/core/drlibc/drlibc.c
@@ -59,18 +59,6 @@ safe_read_if_fast(const void *base, size_t size, void *out_buf)
     return d_r_safe_read(base, size, out_buf);
 }
 
-WEAK void
-d_r_internal_error(const char *file, int line, const char *expr)
-{
-    /* Do nothing for non-core. */
-}
-
-WEAK bool
-ignore_assert(const char *assert_file_line, const char *expr)
-{
-    return true;
-}
-
 WEAK int
 d_r_strcmp(const char *left, const char *right)
 {
@@ -308,4 +296,15 @@ find_script_interpreter(OUT script_interpreter_t *result, IN const char *fname,
     result->argv[argc] = NULL;
     return true;
 }
+#endif
+
+#ifdef WINDOWS
+/***************************************************************************
+ * Windows mode switching support.
+ */
+
+/* We set a default equal to the observed 0x2b value on every Windows version.
+ * The core calls d_r_set_ss_selector() to update to the underlying value.
+ */
+int d_r_ss_value = 0x2b;
 #endif

--- a/core/drlibc/drlibc.h
+++ b/core/drlibc/drlibc.h
@@ -1,5 +1,5 @@
 /* **********************************************************
- * Copyright (c) 2011-2019 Google, Inc.  All rights reserved.
+ * Copyright (c) 2011-2020 Google, Inc.  All rights reserved.
  * Copyright (c) 2000-2010 VMware, Inc.  All rights reserved.
  * **********************************************************/
 
@@ -124,5 +124,13 @@ bool
 find_script_interpreter(OUT script_interpreter_t *result, IN const char *fname,
                         ssize_t (*reader)(const char *pathname, void *buf, size_t count));
 #endif /* UNIX */
+
+#if defined(WINDOWS) && !defined(X64)
+/* Meant to be called at initialization time when .data is writable and races are
+ * not a concern.
+ */
+void
+d_r_set_ss_selector();
+#endif
 
 #endif /* _DR_LIBC_H_ */

--- a/core/drlibc/drlibc_notdr_error.c
+++ b/core/drlibc/drlibc_notdr_error.c
@@ -1,0 +1,45 @@
+/* **********************************************************
+ * Copyright (c) 2020 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Default implementation to avoid the user of drlibc having to supply one.
+ * We need to separate these as the MSVC linker will not complain about duplicate
+ * symbols if an .obj has just one symbol in it.
+ */
+
+#define DR_NO_FAST_IR /* Avoid pulling in deps from instr_inline.h from globals.h */
+#include "../globals.h"
+
+WEAK void
+d_r_internal_error(const char *file, int line, const char *expr)
+{
+    /* Do nothing for non-core. */
+}

--- a/core/drlibc/drlibc_notdr_ignoreassert.c
+++ b/core/drlibc/drlibc_notdr_ignoreassert.c
@@ -1,0 +1,45 @@
+/* **********************************************************
+ * Copyright (c) 2020 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Default implementation to avoid the user of drlibc having to supply one.
+ * We need to separate these as the MSVC linker will not complain about duplicate
+ * symbols if an .obj has just one symbol in it.
+ */
+
+#define DR_NO_FAST_IR /* Avoid pulling in deps from instr_inline.h from globals.h */
+#include "../globals.h"
+
+WEAK bool
+ignore_assert(const char *assert_file_line, const char *expr)
+{
+    return true;
+}

--- a/core/drlibc/drlibc_x86.asm
+++ b/core/drlibc/drlibc_x86.asm
@@ -517,6 +517,19 @@ sml_return_to_32:
         END_FUNC(FUNCNAME)
 
 /*
+ * void d_r_set_ss_selector()
+ */
+DECL_EXTERN(d_r_ss_value)
+# undef FUNCNAME
+# define FUNCNAME d_r_set_ss_selector
+        DECLARE_FUNC(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+        mov      eax, ss
+        mov      DWORD SYMREF(d_r_ss_value), eax
+        ret
+        END_FUNC(FUNCNAME)
+
+/*
  * int switch_modes_and_call(uint64 func, void *arg1, void *arg2, void *arg3)
  */
 # undef FUNCNAME
@@ -564,6 +577,15 @@ smc_transfer_to_64:
         jmp      fword ptr [esp]
 smc_return_to_32:
         add      esp, 8          /* clean up far jmp target */
+        /* i#4091: Work around an AMD processor bug where after switching from 64-bit
+         * back to 32-bit, if a thread switch happens around the same time, the
+         * SS segment descriptor gets corrupted somehow and any ESP reference
+         * raises an access violation with an undocumented Parameter[0]=00000003.
+         * Re-instating the proper descriptor by re-loading the selector seems
+         * to solve the problem.
+         */
+        mov      eax, DWORD SYMREF(d_r_ss_value)
+        mov      ss, eax
         pop      ebx             /* restore callee-saved reg */
         ret                      /* return value already in eax */
         END_FUNC(FUNCNAME)

--- a/core/drlibc/drlibc_x86.asm
+++ b/core/drlibc/drlibc_x86.asm
@@ -584,8 +584,8 @@ smc_return_to_32:
          * Re-instating the proper descriptor by re-loading the selector seems
          * to solve the problem.
          */
-        mov      eax, DWORD SYMREF(d_r_ss_value)
-        mov      ss, eax
+        mov      ebx, DWORD SYMREF(d_r_ss_value)
+        mov      ss, ebx
         pop      ebx             /* restore callee-saved reg */
         ret                      /* return value already in eax */
         END_FUNC(FUNCNAME)

--- a/core/win32/os.c
+++ b/core/win32/os.c
@@ -1895,6 +1895,9 @@ os_take_over_init(void)
         GLOBAL_DCONTEXT, INIT_HTABLE_SIZE_TAKEOVER,
         80 /* load factor: not perf-critical */, HASHTABLE_SHARED | HASHTABLE_PERSISTENT,
         takeover_table_entry_free _IF_DEBUG("takeover table"));
+#    ifndef X64
+    d_r_set_ss_selector();
+#    endif
 }
 
 /* We need to distinguish a thread intercepted via APC hook but that is in ntdll

